### PR TITLE
conv2d decomposition via matmul

### LINKED ISSUE #444: Enable operators conv1d and conv2d
### What feature would you like to be added?

Run torch.conv1d and torch.conv2d on spyre

### Why is this feature needed?

The convolution operator is used by several models that are to be run on Spyre.

### Any alternative solution that you may want to share?

No

### Anything else we need to know?

No

### PR DESCRIPTION
**Pull request by umacdevi**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Provides a decomposition of conv2d using unfold and batch matrix multiplication, matmul

#### Which issue(s) this PR is related to:

https://github.com/torch-spyre/torch-spyre/issues/444

#### Special notes for your reviewer:
Since DeepTools currently does not have conv2d implementations that work on full convolution with large kernels such as the one used in Mistral-Small-3-2-24B-Instruct-2506 to be enabled for 2Q, this workaround implementation via matmul is provided.

#### Additional note:
The following script can be used as a template for testing.

```
import torch
from torch_spyre._C import get_device_dtype, SpyreTensorLayout

torch.manual_seed(0xAFFE)

x = torch.randn(8, 64, 128, 128, dtype=torch.float16)
y = torch.randn(64, 64, 3, 3, dtype=torch.float16)
z = torch.ones(64, dtype=torch.float16)

x_dev = x.to("spyre")
y_dev = y.to("spyre")
z_dev = z.to("spyre")

cpu = torch.conv2d(x, y, z, padding=(1,1), groups=1)

aiu = torch.compile(torch.conv2d)(x_dev, y_dev, z_dev, padding=(1,1), groups=1).cpu()

print(f"Shapes: cpu: {cpu.shape}, aiu: {aiu.shape}")
print(cpu[0, 0])
print(aiu[0, 0])

print(torch.abs(cpu - aiu).amax())
```

x and y tensors can be replaced with the following tensors from the mistral model.
```
x = torch.randn(1, 3, 392, 532, dtype=torch.float16)
y = torch.randn(1024, 3, 14, 14, dtype=torch.float16)
```

padding, groups, and stride can be modified as well.

### CODE REVIEW

File: torch_spyre/_inductor/decompositions.py [hunk 1]
```diff
@@ -672,6 +672,91 @@ def bitwise_and(input1: torch.Tensor, input2: torch.Tensor) -> torch.Tensor:
...
+        raise Unsupported(f"conv2d_via_bmm: expected 4D input, got {input.dim()}D")
+
+    N, C_in, H_in, W_in = input.shape
+    C_out, C_in_per_group, K_h, K_w = weight.shape
+
+    stride_h, stride_w = stride[0], stride[1]
+    pad_h, pad_w = padding[0], padding[1]
+    dil_h, dil_w = dilation[0], dilation[1]
+
+    assert C_in == groups * C_in_per_group, (
```
Comment (dgrove-oss) [inline torch_spyre/_inductor/decompositions.py]: For consistency, shouldn't this also be a check + raise, same as the other checks above?

File: torch_spyre/_inductor/customops.py [hunk 2]
```diff
@@ -362,6 +363,92 @@ def _(input: torch.Tensor, dim: int, keepdim: bool = False):
...
+    When reshaping produces a shape with innermost dimension that doesn't align
+    with stick boundaries (64 elements for fp16), the Inductor coordinate
+    computation fails. This op moves to CPU, reshapes, then moves back to Spyre.
+
+    This is similar to unfold, which also uses CPU fallback for correct layouts.
+    """
+    warn_fallback("torch.ops.spyre.reshape_via_cpu")
+    input_cpu = input.to("cpu")
+    result_cpu = input_cpu.reshape(shape)
+    return result_cpu.to(input.device)
```
Comment (dgrove-oss) [inline torch_spyre/_inductor/customops.py:440]: The fallback will cause us to split the graph into multiple SuperDSC bundles.  This in turn will block working set reduction from being applied across conv2D operations.

If they don't exist already, please open issues to track implementation of on-device versions of these operations that are being done via fallback in this PR.

**Review by dgrove-oss — COMMENTED** [review]
The PR should include test cases for the new operation.

No objection to merging once a few test cases are added.   As noted in the detailed comments, the cpu_fallback is going to act as a barrier to meaningful optimization. So this approach is ok for basic enablement, but we will need to do the layout operations on device to get a performant version.

File: torch_spyre/_inductor/customops.py (hunk 2 above)
Comment (umacdevi) [reply torch_spyre/_inductor/customops.py:440]: Actually, all the reshape_via_cpu() calls issued from conv2d_via_bmm_decomp() can be performed before the graph with matmul is executed on spyre. So no graph breaks are introduced because of reshapes used within the conv2d decomposition function.

File: torch_spyre/_inductor/decompositions.py (hunk 1 above)
Comment (umacdevi) [reply torch_spyre/_inductor/decompositions.py]: Changed the assert to condition check + raise.

**Comment by umacdevi** [comment]
Added a few tests to tests/test_inductor_ops.py.

The resize_via_cpu() calls do not seem to be inducing graph breaks, yet opening an issue to track systematically and close.

**Comment by dgrove-oss** [comment]
@umacdevi -- could you resolve the merge conflicts so we can proceeding with merging this?  Thanks

**Comment by umacdevi** [comment]
> @umacdevi -- could you resolve the merge conflicts so we can proceeding with merging this? Thanks

Resolved.
^^^ @dgrove-oss

**Review by dgrove-oss — DISMISSED** [review]
lgtm

**Review by dgrove-oss — APPROVED** [review]
